### ISSUE TITLE
Bug in find_nullclines

### ISSUE THREAD

**Issue by dionperd** [robclewley/pydstool#163]
Dear developers,

I am getting an error upon trying to find nullclines of the model below, although I can integrate it without problems.

The error is:

TypeError: _auxfn_fpre() takes 3 positional arguments but 5 were given

where fpre is an auxiliary function.

The model contains an if() function which also has a max() function inside it. I don't know if this is the root of the problem.

I have tried to also remove the third state variable by converting it into a parameter, but I got a similar error, complaining for 4 positional arguments instead of 3.

I work on a Mac 11.4, and python version 3.7.10.

You can find more details below, and a .txt with my code attached.
[code.txt](https://github.com/robclewley/pydstool/files/6809303/code.txt)

I would be grateful for any help with this bug, since I haven't managed to debug it myself!

Thank you very much in advance.

D.P.

The error in details:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-101c9de1cc0a> in <module>
      3 # The fixed point found is also provided to help locate the nullclines.
      4 nulls_V, nulls_U = pp.find_nullclines(ode, 'V', 'U', n=3, eps=1e-8,  # subdomain=subdomain,
----> 5                                       max_step=0.1, fps=[fp_coord[0]])  # 
      6 
      7 

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/Toolbox/phaseplane.py in find_nullclines(gen, xname, yname, subdomain, fps, n, t, eps, crop_tol_pc, jac, max_step, max_num_points, only_var, seed_points, strict_domains, pycont_cache)
    982             while not done:
    983                 try:
--> 984                     P_y['null_curve_y'].forward()
    985                 except PyDSTool_ExistError:
    986                     tries += 1

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/PyCont/Continuation.py in forward(self)
   1243         self.CurveInfo = PointInfo()
   1244         if self.sol is None:
-> 1245             self._compute(v0=self.initdirec)
   1246             self.sol = self._curveToPointset()
   1247 

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/PyCont/Continuation.py in _compute(self, x0, v0, direc)
   1008         if self.sysfunc == self.CorrFunc:
   1009             # It's codim 1 and can use the existing Jac (in case it's symbolic)
-> 1010             J = self.CorrFunc.jac(x0)
   1011             J_coords = J[:,self.coords]
   1012             J_params = J[:,self.params]

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/PyCont/TestFunc.py in __call__(self, *cargs)
     84 
     85     def __call__(self, *cargs):
---> 86         self.lastval = self.func(*cargs)
     87         return self.lastval
     88 

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/PyCont/TestFunc.py in diff(self, x0, ind)
    102             ei = zeros(self.n, float)
    103             ei[ind[i]] = 1.0
--> 104             J[:,i] = (self.func(x0+eps*ei)-self.func(x0-eps*ei))/(2*eps)
    105         return J
    106 

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/PyCont/Continuation.py in _system(self, X)
    453             # autonomous system, t doesn't matter
    454             t = 0
--> 455         return self.gensys.Rhs(t, VARS, self.parsdict, asarray=True)[self.varsindices]
    456 
    457 

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/Generator/Vode_ODEsystem.py in Rhs(self, t, xdict, pdict, asarray)
    898             p = sortedDictValues(self._FScompatibleNames(pdict))
    899         i = _pollInputs(sortedDictValues(self.inputs), t, self.checklevel)
--> 900         return getattr(self, self.funcspec.spec[1])(*[t, x, p + i])
    901 
    902     def Jacobian(self, t, xdict, pdict=None, asarray=True):

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/Generator/ODEsystem.py in _specfn(ds, t, x, parsinps)

~/opt/anaconda3/lib/python3.7/site-packages/PyDSTool-0.91.0-py3.7.egg/PyDSTool/Generator/ODEsystem.py in _auxfn_fpost(ds, parsinps, U)

TypeError: _auxfn_fpre() takes 3 positional arguments but 5 were given

**Comment by robclewley** [comment]
Thanks for reporting and apologies for the cryptic error message. It's a deep-down issue with the nullcline prep where it fails to automatically adjust the correct use of call arguments in any auxiliary functions that call other aux functions. In PyDSTool, the aux function call arguments only refer to dependent variables. When any of those variables become fixed for the nullcline system, as is the case here, they become system parameters and will be passed in automatically. Meanwhile, they must be removed from the argument list of any function call instances. When they are not removed, the receiving functions will get too many arguments passed. I think I know how to fix it and I'll take a look soon.

**Comment by dionperd** [comment]
Thanks a lot for you quick reply!

I am leaving on vacation on Monday for about 10 days. 
So, no rush!

At the same time, I am trying to do some bifurcation diagrams on a simple model, and I am getting into some trouble with non-convergence even when I set min step size as low as 1e-10, and although the model has no discontinuity, I proved the Jacobian etc. I will research a bit more to see if it is a science issue, or some wrong use of the tool from my side.

If I don't manage, I might make an issue about it as well.

Many thanks for a great tool, and the support!
